### DISCUSSION TITLE
Server span is not getting closed when using embedded Tomcat v9 with SpringBoot v2.6.15

### DISCUSSION THREAD

**Discussion by MaheshGPai** [open-telemetry/opentelemetry-java-instrumentation#10730] [Q&A] [answered]
I see that the auto instrumentation code is not closing the Server span when application is deployed with Tomcat v9 with SpringBoot v2.6.15

The reason for this is the that the below condition evaluates to false when the Request processing ends:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/b73901047b63b392fc2542ab962f2e04dce02843/instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatHelper.java#L52
```
    if (throwable != null || mustEndOnHandlerMethodExit(request)) {
      instrumenter.end(context, request, response, throwable);
    }
```
On further debugging the issue seems to be the HttpServletRequest type. The instrumentation is expecting an instance of `jakarta.servlet.http.HttpServletRequest`. But the actual instance type is `javax.servlet.http.HttpServletRequest`

Due to this mis-match, the servletEntityProvider returns null @ https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/b73901047b63b392fc2542ab962f2e04dce02843/instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatHelper.java#L58

```
  private boolean mustEndOnHandlerMethodExit(Request request) {
    REQUEST servletRequest = servletEntityProvider.getServletRequest(request);
    return servletRequest != null && servletHelper.mustEndOnHandlerMethodExit(servletRequest);
  }
```

Can you please advice how to over come this issue?

**Comment by laurit** [comment]
are you bundling both jakarta and javax servlet api inside your application?

**Reply by MaheshGPai** [reply]
I looked at the dependencies. `jarkarta-servlet-api` is indeed getting pulled in by a dependent library. I'll have to try and exclude that.

**Reply by MaheshGPai** [reply]
Even with excluding `jaraka-servlet-api` I'm still seeing the same behavior

**Reply by laurit** [reply]
Would it be possible to provide a minimal sample application that reproduces this issue?

**Reply by MaheshGPai** [reply]
There was one more jakarta dependecy that was causing this behaviour. After execluding `jetty-jakarta-servlet-api`, I see the spans are getting closed and exported as expected.

**Reply by MaheshGPai** [reply]
Is there any other way to handle? I presume this mixing up of jarkata & javax modules are quite common?

**Reply by laurit** [reply, marked as answer]
> I presume this mixing up of jarkata & javax modules are quite common?

I don't think it is common as both of these dependencies can not reasonably co-exist in the same application. If both of them are included then it very likely that it is because of a mistake.

I have create a PR that hopefully resolves this https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10757

**Reply by MaheshGPai** [reply]
Fair point. Thanks a lot for the pointers and the PR. Helped in quickly solving this!
